### PR TITLE
Rework Object searching to behave more consistently

### PR DESCRIPTION
**Pull request by InspectorCaracal**
#### Brief overview of PR changes/additions
This PR makes modifications to both the DefaultObject manager methods of `get_objs_by_key_and_alias` and `search_object`.

`get_objs_by_key_and_alias` will now properly respect partial matches in its queries, leading to the expected result of an empty search string partial-matching all candidates and additionally running fewer queries.

`search_object` has been moderately streamlined, removing the need for special-case handling of tags or search terms, and applying all filtering keywords consistently to all types of searches rather than unintuitively picking and choosing based on which combination of keywords were provided.

#### Motivation for adding to Evennia
Usability improvements

#### Other info (issues closed, discussion etc)
The undesirable behavior brought up on Discord where searching with tag criteria but no search string is resolved by this PR. It additionally resolves several equally problematic behaviors that currently result when giving the search method an empty search string, such as returning an empty result when people typically expect an empty string to partial-match *all* keys not none, and the unfortunate tag-only search disregarding all other keywords entirely.

### CODE REVIEW

**Comment by InspectorCaracal** [comment]
Whoops.

Converting this to a draft because a) there's more failing unit tests, and b) I just realized while working on adding new tests for the two methods I modified that I unwittingly lost the match scoring mechanism. Since that needs to be accounted for, I need to rethink some of my changes to `get_objs_with_key_or_alias`

**Comment by InspectorCaracal** [comment]
On further review of how `string_partial_matching` works, it seems that the "score" is just how many of the input words were matched in the alternative text. Since the score is zeroed when any input words are lacking, this means my single query will in practice return equivalent results since it requires that all input words are partially matched in order. As such, I'm leaving the modifications as I originally wrote them.

The failed unit test is a bit more of a pain, however. It seems as though the clothing contrib, rather than having a proper usage feedback on the `remove` command when used with no args, was set up to verify that it returned a "no object found" error instead. Since caller.search is fuzzy by default, this is obviously failing now that the empty term fuzzy search works "correctly"....

File: evennia/contrib/game_systems/clothing/tests.py
Comment (InspectorCaracal) [inline evennia/contrib/game_systems/clothing/tests.py:88]: I've temporarily just commented out this assertion since I intend to make a second PR to correct the oversight in the contrib command, but I'm unsure if this is the correct approach.

I could alternatively write the other PR, wait for it to be merged, and rebase this one onto it.

Or, I could leave this one failing with the assumption I'll correct it. (I'm not a fan of this third option. >.>)

**Review by Griatch — COMMENTED** [review]
At first glance it looks like it would comprehensive. I'm not entirely convinced that it's obvious that you'd find all candidates by fuzzy-matching with an empty string honestly, but it does simplify things a bit too. Also nice use of iregex. 

I need to run this more to check before merging this.

**Comment by InspectorCaracal** [comment]
> I'm not entirely convinced that it's obvious that you'd find all candidates by fuzzy-matching with an empty string honestly, but it does simplify things a bit too. 

I tried it out in a test game and it seemed like it was working, so let me know if you find cases where it doesn't. 👍

File: evennia/objects/manager.py
Comment (Griatch) [inline evennia/objects/manager.py:486]: This first `if` conditional seems like it is superfluous ... both this and the `else:` statement fulfills the same condition (being outside of the range).

**Review by Griatch — CHANGES_REQUESTED** [review]
Testing some more, I think this is a nice change, it also cleans up/shortens the code a lot, that `iregex=` is a nice touch. Apparently I wrote much of this code a whopping 12 years ago! :astonished: 

(By the way, changing the search result in a way that breaks unit tests makes it clear this is a backwards-incompatible change.)  

So far it seems to behave as expected in my testing. I comment on some code below. I did find one piece of strangeness though: 

```python
> py self.search("")
Griatch-1
Limbo-2
rose-3
cactus-5
box-1
box-2
box-3
```
Whereas the box-1..3 are the expected multi-match differentiators, the other added numbers appear for _single_ matches and appear to be the dbrefs rather than the run-numbers. Expectedly, doing e.g. `py self.search("rose-3")` will fail. As far as I can tell, this PR interacts in an unexpected way with the listing display of multi-matches.

**Comment by InspectorCaracal** [comment]
> Whereas the box-1..3 are the expected multi-match differentiators, the other added numbers appear for single matches and appear to be the dbrefs rather than the run-numbers.

Ohh, that's weird and very interesting. I'm looking forward to figuring that one out!

**Comment by InspectorCaracal** [comment]
Interestingly, in a fresh gamedir on this branch of evennia, I had a completely different result:
```
> py self.search('')
>>> self.search('')
More than one match for '' (please narrow target):
 admin-1
 Limbo-2
 box-3
 box-4
 box-5
 ```

This is also technically wrong, but it's more in line with what I would've expected. :thinking:

It seems to also actually be highlighting a flaw in the existing multimatching system's ability to handle partial matches - if I switch to the current evennia main, it behaves like this:
```
>>> self.search('red')
More than one match for 'red' (please narrow target):
 red box-1
 red block-2
>>> self.search('red block-2')
Could not find 'red block-2'.
```

I'll try a few more times to see if I can replicate your weird result, but I'll make a separate bug report for the partial-match multimatching issue.

**Comment by InspectorCaracal** [comment]
@Griatch I officially cannot replicate anything like your result. I've tried multiple different collections of items, quelled and unquelled on the superuser and auto-created character, destroying and creating new objects, and even made a command that does nothing but an empty-string search to try it on a Player-level character, and the results are *always* numbered in simple list order.

```
More than one match for '' (please narrow target):
 box-1
 box-2
 red box-3
 blue box-4
 red block-5
 red block-6
 field-7
 south-8 [s]
 tester-9
 box-10
 ```
 
 So as far as I can tell, that wasn't/isn't an issue with this PR.

File: evennia/objects/tests.py
Comment (Griatch) [inline evennia/objects/tests.py:253]: This is done twice? A typo or some reason?

File: evennia/objects/tests.py
Comment (Griatch) [inline evennia/objects/tests.py:257]: Should this be `self.assertNone` to be explicit?

File: evennia/objects/manager.py
Comment (Griatch) [reply evennia/objects/manager.py:486]: This still seems to be the case.

File: evennia/contrib/game_systems/clothing/tests.py
Comment (Griatch) [reply evennia/contrib/game_systems/clothing/tests.py:88]: It would be a good idea to make sure all unit tests work correctly so we don't merge a version that is actually broken for some use cases (even if the version is short-lived). So I think it makes sense to fix this contrib/unit test as part of this PR as well.

**Review by Griatch — CHANGES_REQUESTED** [review]
:thinking: I can't reproduce the separate box-1,2,3 behavior I saw now either. It does make me a little worried :sweat_smile:

Overall, partial matching seems to be working fine with this change. As you mention, this does highlight a problem with multimatch display for searching for `""` though - the resulting list of multimatches is nonsense and not actionable (this is not a regression from this PR though, as you say, I see the same behavior in main). That could be a separate investigation.

I do think this should not merge until it actually comes with a fully updated unit test suite though (rather than commented out tests); those adjustments can/should be done in the same PR IMO. I also give some comments on the unit tests for this change.

**Comment by InspectorCaracal** [comment]
> I do think this should not merge until it actually comes with a fully updated unit test suite though (rather than commented out tests); those adjustments can/should be done in the same PR IMO. I also give some comments on the unit tests for this change.

Thanks! I was waiting for an answer to this before tackling the needed changes so I could do everything at once.

File: evennia/objects/tests.py
Comment (InspectorCaracal) [reply evennia/objects/tests.py:257]: If we want explicit, then I'll need to check if it's equal to an empty list. I can do that. 👍

**Comment by Griatch** [comment]
Ping, not much left to do on this one :)

**Comment by InspectorCaracal** [comment]
Haven't forgotten about this one either, hopefully will be wrapping it up soon.

File: evennia/objects/manager.py
Comment (InspectorCaracal) [reply evennia/objects/manager.py:486]: The outer `if match_number is not None` is functionally relevant. In terms of code logic, it means that the inner checks regarding whether the match-number index is valid will only be checked if there is, in fact, a match-number provided in the term to begin with. I don't believe this to be superfluous or redundant.

Comment (InspectorCaracal) [reply evennia/objects/manager.py:486]: Ah, I've been misinterpreting what you meant by "the first if"; since it's below two ifs I thought you meant the first of the two you commended below.

**Review by Griatch — APPROVED** [review]
Thanks for your work addressing the comments. Looks good now, and with the fixed unit test too.

**Pull request merged**
